# Naive Bayes Case Study on Titanic Dataset


In [1]:
import pandas as pd
data=pd.read_csv("C:\\Users\\Ruchi Goyal\\Downloads\\titanic.csv")
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [2]:
data.shape

(891, 12)

In [4]:
data1=data.drop(["PassengerId","Name","SibSp","Parch","Ticket","Cabin"],axis=1)
data1.head()

Survived  Pclass     Sex   Age     Fare Embarked
0         0       3    male  22.0   7.2500        S
1         1       1  female  38.0  71.2833        C
2         1       3  female  26.0   7.9250        S
3         1       1  female  35.0  53.1000        S
4         0       3    male  35.0   8.0500        S

In [5]:
data1["Embarked"].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [6]:
data1["Sex"]=data1["Sex"].map({"male":0,"female":1})
data1["Embarked"]=data1["Embarked"].map({"S":0,"C":1,"Q":2})
data1.head()

Survived  Pclass  Sex   Age     Fare  Embarked
0         0       3    0  22.0   7.2500       0.0
1         1       1    1  38.0  71.2833       1.0
2         1       3    1  26.0   7.9250       0.0
3         1       1    1  35.0  53.1000       0.0
4         0       3    0  35.0   8.0500       0.0

In [7]:
data1.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
Fare          0
Embarked      2
dtype: int64

In [9]:
X = data1.drop(["Survived","Embarked"], axis=1)
y = data1["Survived"]
X.head()

Pclass  Sex   Age     Fare
0       3    0  22.0   7.2500
1       1    1  38.0  71.2833
2       3    1  26.0   7.9250
3       1    1  35.0  53.1000
4       3    0  35.0   8.0500

In [10]:
from sklearn.impute import SimpleImputer 
imputer = SimpleImputer(strategy='mean')  
X_imputed = imputer.fit_transform(X)
X_imputed[:10]


array([[ 3.        ,  0.        , 22.        ,  7.25      ],
       [ 1.        ,  1.        , 38.        , 71.2833    ],
       [ 3.        ,  1.        , 26.        ,  7.925     ],
       [ 1.        ,  1.        , 35.        , 53.1       ],
       [ 3.        ,  0.        , 35.        ,  8.05      ],
       [ 3.        ,  0.        , 29.69911765,  8.4583    ],
       [ 1.        ,  0.        , 54.        , 51.8625    ],
       [ 3.        ,  0.        ,  2.        , 21.075     ],
       [ 3.        ,  1.        , 27.        , 11.1333    ],
       [ 2.        ,  1.        , 14.        , 30.0708    ]])

In [11]:
import numpy as np
np.isnan(X_imputed).any()

False

In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.3, random_state=42)

In [13]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
model.fit(X_train, y_train)

GaussianNB()

In [14]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
y_pred = model.predict(X_test)

print("Accuracy score:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:", confusion_matrix(y_test, y_pred))
print("Classification Report:", classification_report(y_test, y_pred))

Accuracy score: 0.7761194029850746
Confusion Matrix: [[126  31]
 [ 29  82]]
Classification Report:               precision    recall  f1-score   support

           0       0.81      0.80      0.81       157
           1       0.73      0.74      0.73       111

    accuracy                           0.78       268
   macro avg       0.77      0.77      0.77       268
weighted avg       0.78      0.78      0.78       268



## Summary

### Out of 268 predictions, around 78% predictions are correct which is a good accuracy score.

### Some room for improvement, especially in distinguishing class 1



## True Negatives (TN = 126): Out of 268, 126 correctly predicted class 0

## False Positives (FP = 31): Out of 268, 31 Class 0 misclassified as class 1

## False Negatives (FN = 29): Out of 268, 29 Class 1 misclassified as class 0

## True Positives (TP = 82): Out of 268, 82 Correctly predicted class 1